trie/zk_merkle_tree.go

```diff
@@ -0,0 +1,152 @@
+	byte32Array := bytesToByte32Array(value)
+	return z.MerkleTree.TryUpdate(panicIfLenNot20or32(key), uint32(len(byte32Array)), byte32Array)
+}
+
+func (z *ZkMerkleTree) TryUpdateAccount(address common.Address, account *types.StateAccount) error {
+	panicIfLenNot20or32(address.Bytes())
+	value, flag := account.MarshalFields()
+	return z.MerkleTree.TryUpdate(address.Bytes(), flag, value)
+}
+
+func (z *ZkMerkleTree) TryDelete(key []byte) error {
+	return z.MerkleTree.TryDelete(key)
+}
+
+func (z *ZkMerkleTree) TryDeleteAccount(address common.Address) error {
+	return z.MerkleTree.TryDelete(address.Bytes())
+}
+
+func (z *ZkMerkleTree) NodeIterator(startKey []byte) NodeIterator {
+	return newZkMerkleTreeIterator(z.Hash(), z.db.Get, startKey)
+}
+
+func (z *ZkMerkleTree) Hash() common.Hash { return common.BytesToHash(z.MerkleTree.Hash()) }
+
+func (z *ZkMerkleTree) Commit(collectLeaf bool) (common.Hash, *NodeSet) {
+	err := z.ComputeAllNodeHash(func(node zk.TreeNode) error { return z.db.Put(node.Hash()[:], node.CanonicalValue()) })
+	if err != nil {
+		log.Error(fmt.Sprintf("fail to commit zk merkle trie. %v", err))
+	}
+	rawdb.WriteLegacyTrieNode(z.db.db.diskdb, z.Hash(), z.RootNode().Value())
+	// Since NodeSet relies directly on mpt, we can't create a NodeSet.
+	// Of course, we might be able to force-fit it by implementing the node interface.
+	// However, NodeSet has been improved in geth, it could be improved to return a NodeSet when a commit is applied.
+	// So let's not do that for the time being.
+	// related geth commit : https://github.com/ethereum/go-ethereum/commit/bbcb5ea37b31c48a59f9aa5fad3fd22233c8a2ae
+	return z.Hash(), nil
+}
+
+func (z *ZkMerkleTree) Prove(key []byte, fromLevel uint, proofDb ethdb.KeyValueWriter) error {
+	err := z.MerkleTree.Prove(key, fromLevel, func(node zk.TreeNode) error {
+		if leaf, ok := node.(*zk.LeafNode); ok {
+			preImage := z.GetKey(leaf.KeyHash.Bytes())
+			if len(preImage) > 0 {
+				leaf.KeyPreimage = &zkt.Byte32{}
+				copy(leaf.KeyPreimage[:], preImage)
+			}
+		}
+		return proofDb.Put(node.Hash()[:], node.Value())
+	})
+	if err != nil {
+		return err
+	}
+	// we put this special kv pair in db so we can distinguish the type and make suitable Proof
+	return proofDb.Put(magicHash, zktrie.ProofMagicBytes())
+}
+
+func (z *ZkMerkleTree) Copy() MerkleTrie { panic("implement me") }
+
+func (z *ZkMerkleTree) IsZk() bool { return true }
+
+func panicIfLenNot20or32(b []byte) []byte {
+	if len(b) != 32 && len(b) != 20 {
+		panic(fmt.Errorf("do not support length except for 120bit and 256bit now. data: %v len: %v", b, len(b)))
+	}
+	return b
+}
+
+func bytesToByte32Array(bytes []byte) []zkt.Byte32 {
+	byte32Values := *new([]zkt.Byte32)
+	for offset := 0; offset < len(bytes); offset += 32 {
+		end := math.Min(float64(offset+32), float64(len(bytes)))
+		byte32Values = append(byte32Values, *zkt.NewByte32FromBytes(bytes[offset:int(end)]))
+	}
+	return byte32Values
+}
```

Comment:
This performs slightly better! And It would be better to copy the values to `zkt.Bytes32` directly avoiding heap allocation.

```suggestion
func bytesToByte32Array(bytes []byte) []zkt.Byte32 {
	byte32Values := make([]zkt.Byte32, (len(bytes)+31)/32)
	offset := 0
	for i := range byte32Values {
		end := math.Min(float64(offset+32), float64(len(bytes)))
		byte32Values[i] = *zkt.NewByte32FromBytes(bytes[offset:int(end)])
		offset += 32
	}
	return byte32Values
}
```